src/tss2-fapi/ifapi_eventlog_system.c

```diff
@@ -0,0 +1,766 @@
+ * to an int64 number.
+ *
+ * param[in] token the string representing the number.
+ * param[out] num the converted number.
+ * @retval true if token represents a number
+ * @retval false if token does not represent a number.
+ */
+static bool
+get_number(const char *token, int64_t *num)
+{
+    int itoken = 0;
+    int pos = 0;
+    if (strncmp(token, "0x", 2) == 0) {
+        itoken = 2;
+        sscanf(&token[itoken], "%"PRIx64"%n", num, &pos);
+    } else {
+        sscanf(&token[itoken], "%"PRId64"%n", num, &pos);
+    }
+    if ((size_t)pos == strlen(token) - itoken)
+        return true;
+    else
+        return false;
+}
+
+/** Get the index of a sub string after a certain prefix.
+ *
+ * The prefixes from table tss_const_prefixes will be used for case
+ * insensitive comparison.
+ *
+ * param[in] token the token with a potential prefix.
+ * @retval the position of the sub string after the prefix.
+ * @retval 0 if no prefix is found.
+ */
+static int
+get_token_start_idx(const char *token)
+{
+    int itoken = 0;
+    char *entry;
+    int i;
+
+    for (i = 0, entry = tss_const_prefixes[0]; entry != NULL;
+            i++, entry = tss_const_prefixes[i]) {
+        if (strncasecmp(token, entry, strlen(entry)) == 0) {
+            itoken += strlen(entry);
+            break;
+        }
+    }
+    return itoken;
+}
+
+bool digest2_accumulator_callback(TCG_DIGEST2 const *digest, size_t size,
+                                  void *data){
+    if (digest == NULL || data == NULL) {
+        LOG_ERROR("neither parameter may be NULL");
+        return false;
+    }
+    size_t *accumulator = (size_t*)data;
+
+    *accumulator += sizeof(*digest) + size;
+
+    return true;
+}
+/*
+ * Invoke callback function for each TCG_DIGEST2 structure in the provided
+ * TCG_EVENT_HEADER2. The callback function is only invoked if this function
+ * is first able to determine that the provided buffer is large enough to
+ * hold the digest. The size of the digest is passed to the callback in the
+ * 'size' parameter.
+ */
+bool foreach_digest2(
+    tpm2_eventlog_context *ctx,
+    UINT32 event_type,
+    unsigned pcr_index,
+    TCG_DIGEST2 const *digest,
+    size_t count,
+    size_t size) {
+
+    if (digest == NULL) {
+        LOG_ERROR("digest cannot be NULL");
+        return false;
+    }
+
+    bool ret = true;
+    TSS2_RC r;
+
+    size_t i, j;
+    for (i = 0; i < count; ++i) {
+        if (size < sizeof(*digest)) {
+            LOG_ERROR("insufficient size for digest header");
+            return false;
+        }
+
+        const TPMI_ALG_HASH alg = digest->AlgorithmId;
+        const size_t alg_size = ifapi_hash_get_digest_size(alg);
+        if (size < sizeof(*digest) + alg_size) {
+            LOG_ERROR("insufficient size for digest buffer");
+            return false;
+        }
+
+        uint8_t *pcr = NULL;
+        if (pcr_index > TPM2_MAX_PCRS) {
+            LOG_ERROR("PCR%d > max %d", pcr_index, TPM2_MAX_PCRS);
```

Comment:
is this supposed to keep going and not return false immediately?

Reply:
I will add return false.